control wrapping of type signatures

davidchambers opened the pull request:
Signatures are best wrapped before or after:

  - the `::` following the name;
  - the `=>` following the type-class constraints; or
  - a `->` separating argument types.

This wrapping is undesirable:

    get :: Accessible a => TypeRep b -> String -> a -> Maybe
    b

This wrapping is preferable:

    get :: Accessible a => TypeRep b -> String -> a ->
    Maybe b

To achieve the desired wrapping we replace SPACE with NO-BREAK SPACE in positions where wrapping should not occur. For example:

    get :: Accessible_a => TypeRep_b -> String -> a ->
    Maybe_b

Underscores represent U+00A0 in the example above.

I'd :heart: a review from @benperez. I believe `controlWrapping` to be `:ram:`worthy, though I'd appreciate a review of the logic as this project lacks tests.

davidchambers commented on bin/transcribe:
```diff
@@ -16,6 +16,27 @@ var esc = R.pipe(R.replace(/&/g, '&amp;'),
+       R.map(R.split(' => ')),
+       R.map(R.map(R.split(/([(][^()]+[)])/))),
+       R.map(R.map(R.append(''))),
+       R.map(R.map(R.splitEvery(2))),
```
`R.split(pattern, s)` always returns an odd number of strings when `pattern` contains one capturing group, as it does in this case. Given `'(a -> b) -> f a -> f b'`, for example, we'll get `['', '(a -> b)', ' -> f a -> f b']`. We append `''` to give us an even number of strings before pairing. None of the following transformations have any effect on `''`, and it disappears when we `R.join('')` later in the pipeline.

benperez commented on bin/transcribe:
```diff
@@ -16,6 +16,27 @@ var esc = R.pipe(R.replace(/&/g, '&amp;'),
+       R.map(R.map(R.append(''))),
+       R.map(R.map(R.splitEvery(2))),
+       R.map(R.map(R.map(R.over(R.lensIndex(1), nbsp)))),
+       R.map(R.map(R.unnest)),
```
If you're going to `map` then `unnest`, you might as well just `chain`. 

:link:

davidchambers replied:
As a rule, I agree. In this case, though, breaking the rule makes it clear which transformation removes the extra nesting produced by `R.splitEvery(2)`. We have `split, transform, join` rather than `split, transform+join`. I'm happy to include one unnecessary line of code for the clarity it affords.

benperez replied:
Fair enough.

benperez commented:
I love that you were able to make this happen: `R.map(R.map(R.map(R.map(nbsp))))` in real code. One question I have is which of these 2 alternatives is more readable:
```javascript
R.pipe(
  R.map(foo),
  R.map(R.map(bar)),
  R.map(R.map(baz)),
  R.map(R.map(R.map(quux))),
  fizzbuzz
)
```
```javascript
R.pipe(
  R.map(R.pipe(foo,
               R.map(R.pipe(bar,
                            baz,
                            R.map(qux))))),
  fizzbuzz
)
```

davidchambers commented:
> I love that you were able to make this happen: `R.map(R.map(R.map(R.map(nbsp))))` in real code.

:grinning:

@scott-christopher also suggested nested pipes to avoid nested maps. I find nested maps significantly more readable, as the triangular shape of the code makes it easy to see which "joining" transformation undoes each "split". In a sense `R.map(R.map(R.map(R.map(` is just indentation:

```javascript
R.pipe(R.split(' :: '),
             R.split(' => '),
                   R.split(/([(][^()]+[)])/),
                   R.append(''),
                   R.splitEvery(2),
                         R.over(R.lensIndex(1), nbsp),
                   R.unnest,
                         R.split(' -> '),
                               nbsp,
                         R.join(' -> '),
                   R.join(''),
             R.join(' => '),
       R.join(' :: '))
```

| Splitter                    | Joiner                      |
| --------------------------- | --------------------------- |
| `R.split(' :: ')`           | `R.join(' :: ')`            |
| `R.split(' => ')`           | `R.join(' => ')`            |
| `R.split(/([(][^()]+[)])/)` | `R.join('')`                |
| `R.splitEvery(2)`           | `R.unnest`                  |
| `R.split(' -> ')`           | `R.join(' -> ')`            |

benperez commented:
I find your arguments compelling @davidchambers and I award you one :ram: for maniacal yet effective use of ramda in a pull request.

davidchambers commented:
Thank you. I will treasure it always!

davidchambers commented:
I tested this on [Sanctuary][1]. It works an intended. Wrapping can still occur in the *middle* of `->`, but that's a separate issue (which could be addressed by using `→` instead).


[1]: https://github.com/plaid/sanctuary
